staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to slices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// Defines which nodes have access to the devices in the pool.
+	// If the node selector is empty, all nodes have access.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// The generation gets bumped in all slices of a pool whenever device
+	// definitions change. A consumer must only use device definitions from slices
+	// with the highest generation number and ignore all others.
+	PoolGeneration int64 `json:"poolGeneration" protobuf:"bytes,5,name=poolGeneration"`
+
+	// The total number of slices in the pool.
+	// Consumers can use this to check whether they have
+	// seen all slices.
+	PoolSliceCount int64 `json:"poolSliceCount" protobuf:"bytes,6,name=poolSliceCount"`
+
+	// Devices lists all available devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +required
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,7,name=devices"`
+
+	// FUTURE EXTENSION: some other kind of list, should we ever need it.
+	// Old clients seeing an empty Devices field can safely ignore the (to
+	// them) empty pool.
+}
+
+const ResourceSliceMaxSharedCapacity = 128
+const ResourceSliceMaxDevices = 128
+const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// The maximum number of attributes and capacities is 32.
+	//
+	// +optional
+	// +listType=atomic
+	Attributes []DeviceAttribute `json:"attributes,omitempty" protobuf:"bytes,2,rep,name=attributes"`
+
+	// Capacities defines the set of capacities for this device.
+	// The name of each capacity must be unique in that set.
+	//
+	// The maximum number of attributes and capacities is 32.
+	//
+	// +optional
+	// +listType=atomic
+	Capacities []DeviceCapacity `json:"capacities,omitempty" protobuf:"bytes,3,rep,name=capacities"`
```

Comment:
I think we didn't reach a consensus in our prior discussions.

@klueska: what was the future use of having something under "capacities" which isn't a `Quantity`? Can we address that by adding another map?

Reply:
The motivation was to keep the door open for consuming specific elements of a discrete set. With a quantity you can say _how much_ of something you want, but there is no way to say _exactly which ones_ you want.

In one of the prototypes we had, I introduced an `IntRange` abstraction as an example of what this might look like when applied to memory blocks for MIG. In this case however, it wasn't the end-user consuming this capacity, but rather the device itself consuming it from some shared capacity pool.

Which now opens the question -- what's the plan for extending this with support for partitionable devices? Should resources pulled from a shared pool be "implicitly" exposed as part of the capacities that an end-user can consume?  If so, where do we pull them in? If not, why not?

Reply:
The current "capacities" is user-facing. It sound like your `IntRange` is distinct from that, which points towards using `Capacity map[string]resource.Quantity` for the 1.31 API. @thockin: shall we then also use `Attributes map[string]DeviceAttribute` for convenience and consistency?

> what's the plan for extending this with support for partitionable devices? 

That is something that I was hoping to hear from you and John :sweat_smile:. See https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1659249871

Reply:
Even if I don't have an immediate use case, it still feels short-sighted to me to prohibit us from *ever* adding an abstraction that allows one to consume specific items from a discrete set. Can we not make it a `map[string]DeviceCapacity` so that it can be extended similar to `DeviceAttributes` in the future?

Reply:
We can always add new fields to the ResourceSlice as long as it is okay for them to be ignored when not used. More specifically, we can have `Capacity map[string]resource.Quantity` now and add `IntSliceCapacity map[string]StringSlice` later, sharing the same key namespace.

The more interesting question is how those fields appear in a request. We want requests which depend on new features or fields to fail, which can be achieved by making the new types new fields in the `device` object (`device.capacity` now, `device.intSliceCapacity` later) or by simply allowing lookup through the same `device.capacity`. The former leads to compile errors in old clients without `device.intSliceCapacity` support, the latter to runtime errors due or rejection of all devices (depending on whether the CEL expression checks for existence).

Either way, I think `Capacity map[string]resource.Quantity` would not limit future extensions.

Reply:
To me it feels more natural to provide a wrapping struct (i.e. `DeviceCapacity`), knowing we will likely add another field to it someday. It is also more symmetrical with what we do for attributes.

Is there a strong reason *not* to do this, and instead force all capacities to be a `Quantity`?

I don't feel strongly about using a map vs. a list.

Reply:
> knowing we will likely add another field to it someday

I'm not sure about that. Suppose we have a CEL expression which now is `device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi"))`. What are we going to add to `capacities` that is not a quantity?

> Is there a strong reason not to do this, and instead force all capacities to be a Quantity?

I suppose consistency with other "capacity" (yes, singular) fields in other APIs and simplicity (no indirection in the map).

Reply:
4 querstions in this thread:

1) "Capacity" vs "Capacities" - I prefer "Capacity", weakly if we stay list, strinly if we go to a map
2) Map vs list - I think map is better
3) Non-Quantity capacities (user-facing) - see below
4) Non-user-facing capacities - see below

> Suppose we have a CEL expression which now is `device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi"))`

IIUC, Kevin is proposing:

`device.capacities["dra.example.com"].quantities.memory.isGreaterThan(quantity("1Gi"))` or something like that ("it feels more natural to provide a wrapping struct").  I think that would be unfortunate - quantity is going to be the common path, I  think.

> what's the plan for extending this with support for partitionable devices?

I would like to see us patch that on top of this PR, so we see how it fits.  IIRC, we sort of landed on a model where a partitionable (non-leaf) device has "shared" capacity (perhaps "macro-resources") and the leaf devices under it "consume" that macro capacity and "produce" different capacity.  IOW, the leaf-devices (partitions) are congruent to the `Device` defined in this PR, and the macro-device could be schematically different.  Is that right?  

Like, today we might have:

```
Device {
  Name
  Attribs
  Capacity
}
```

and later we might have

```
Device {
  Name
  Attribs
  MacroCapacity
  Partitions[] {
    Name
    Attribs
    MacroConsumed
    Capacity
  }
}
```

The leaf definition is what claim selectors are evaluated against.  Do those selectors need to see how much or which specific macro resources are involved?

If / when we have claims that need to request non-quantity things (and I would push back hard on requesting specific indices, just as I push back on pod requesting specific CPUs :) I think that would be a different field than "capacity".

I think John is out this week, too -- do you have any bandwidth to try to sketch-patch the various partitionable ideas into this?

Reply:
My understanding is that "partitionable devices" don't change the user-facing API at all.

What changes is that selecting a "partition" affects which other "partitions" are still available, so `ResourceSlice` becomes more complex and schedulers need to be aware of that.

*How* `ResourceSlice` needs to change is the open part. I don't think it has anything to do with "capacities".

Reply:
> IIUC, Kevin is proposing:
> 
> device.capacities["dra.example.com"].quantities.memory.isGreaterThan(quantity("1Gi")) or something like that ("it feels more natural to provide a wrapping struct"). 

No, I think he just wanted the ability to extend the ResourceSlice. The `device.capacity["dra.example.com"].field` map then can return either a quantity or "something else", just like attributes. We can achieve that even if we use `Capacity map[string]Quantity` as field in the ResourceSlice, by putting some other capacity field besides it with the restriction that keys must be unique across both.

I am still not sure what this would be used for. It's *not* for partitionable devices, as far as I understand it. @klueska: can you shed some light on what you had in mind when you said "we will likely add another field to it someday"?

As a first step, let's switch to "Capacity" (singular) and maps for *both* capacity and attributes. I prefer map because:
- simplicity
- potential optimization of the CEL evaluation (can directly use map in API type)
- consistency with other `capacity` definitions

> do you have any bandwidth to try to sketch-patch the various partitionable ideas into this?

Yes.

Reply:
> As a first step, let's switch to "Capacity" (singular) and maps for both capacity and attributes.

Included in the new commits that I pushed just now.

Reply:
> 1. "Capacity" vs "Capacities" - I prefer "Capacity", weakly if we stay list, strinly if we go to a map
> 2. Map vs list - I think map is better

Agree on both. For 3 & 4, see below.

> My understanding is that "partitionable devices" don't change the user-facing API at all.

Correct. The fact that "partitionable devices" use shared resources that look a lot like the resources users would consume in for "per-device allocatable resources" is essentially a coincidence (well, it's a result of the real world, but for purposes of the modeling it is a coincidence). What I mean by that is that "partitionable devices" and "per-device allocatable resources" are totally different, unrelated features. They both happen to use some quantity capacities, but they serve different purposes and have different direct users.

I think Kevin wanted "int slice" for "partitionable devices", not for user-facing "per-device allocatable resources". I don't think user-facing selectors at the "leaf" level need to know *anything* about the "macro" provided or consumed. Those are completely internal to how we manage conflicting partitions. 

In this sense, I for sure do not see any need to "wrap" the user-facing capacities. Those are only used in selectors today; later they will be consumable, maybe (if we implement the platform-mediated sharing that is not based on partitioning).

As for the "macro" capacity and consumed, I think since it is not user facing, it's not as critical that we *don't* wrap it. But I still don't think we need to wrap it. In the earlier prototype, I solved the same problem "int range" does by just using different keys. Or we can do what Patrick suggests with respect to a separate field, if we really needed that.

Reply:
Thanks for the confirmation. I had already gone ahead and switched to `Capacity map[QualifiedName]resource.Quantity`, so I think we are done here.

Reply:
I just want to go on the record one last time saying that I think it's short-sighted to assume that all we will ever want here is a simple `Quantity`. I can _maybe_ see wanting to keep it as a `Quantity` in the user-facing API (which we haven't even defined yet), but limiting it here It seems unnecessary (and unwarranted).

I seem to be the only one that feels this way, though, so I will step aside and not argue anymore.

As an aside, John had the notion of consumable resources with a "block size" in one of his early prototypes. If we kept this indirection to a one-of, the user-facing API could still be represented by a `Quantity`, but we would have a natural way of saying in what sized chunks the resources could be consumed.

Reply:
Once we go with a one-of `Device` struct (https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1669911766), changing the type of the `Capacity` field will be possible in a new device type.